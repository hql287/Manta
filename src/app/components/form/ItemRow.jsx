// Libs
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'recompose';

// HOCs
import _withDraggable from './hoc/_withDraggable';

// Styles
import styled from 'styled-components';

const ItemDiv = styled.div`
  position: relative;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex: 1;

  & > div {
    display: flex;
    flex-direction: row;
    margin-right: 10px;
    &:last-child {
      margin-right: 0px;
    }
  }
`;

const ItemDivInput = styled.input`
  min-height: 36px;
  border-radius: 4px;
  padding: 0 10px;
  font-size: 16px;
  display: block;
  width: 100%;
  border: 1px solid #f2f3f4;
  color: #3a3e42;
  font-size: 14px;
`;

const ItemActions = styled.div`
  display: flex !important;
  align-items: center;
  justify-content: center;
  width: 40px;
  margin: 0 !important;
  margin-left: 10px;
`;

const ItemRemoveBtn = styled.a`
  > i {
    color: #ec476e;
  }
`;

// Component
export class ItemRow extends Component {
  constructor(props) {
    super(props);
    this.handleTextInputChange = this.handleTextInputChange.bind(this);
    this.handleNumberInputChange = this.handleNumberInputChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.updateSubtotal = this.updateSubtotal.bind(this);
    this.uploadRowState = this.uploadRowState.bind(this);
    this.removeRow = this.removeRow.bind(this);
  }

  componentWillMount() {
    const { id, description, quantity, price, subtotal } = this.props.item;
    this.setState({
      id,
      description: description || '',
      price: price || '',
      quantity: quantity || '',
      subtotal: subtotal || '',
    });
  }

  handleKeyDown(e) {
    if (e.which === 13) {
      this.props.addItem();
    }
  }

  handleTextInputChange(event) {
    const name = event.target.name;
    const value = event.target.value;
    this.setState({ [name]: value }, () => {
      this.uploadRowState();
    });
  }

  handleNumberInputChange(event) {
    const name = event.target.name;
    const eValue = event.target.value;
    const value = eValue === '' ? '' : parseFloat(eValue);
    this.setState({ [name]: value }, () => {
      this.updateSubtotal();
    });
  }

  updateSubtotal() {
    const currentPrice =
      this.state.price === '' ? 0 : parseFloat(this.state.price);
    const currentQuantity =
      this.state.quantity === '' ? 0 : parseFloat(this.state.quantity);
    let currentSubtotal;
    if (this.state.price === '' || this.state.quantity === '') {
      currentSubtotal = '';
    } else {
      currentSubtotal = currentPrice * currentQuantity;
    }
    this.setState({ subtotal: currentSubtotal }, () => {
      this.uploadRowState();
    });
  }

  uploadRowState() {
    const { updateRow } = this.props;
    updateRow(this.state);
  }

  removeRow() {
    this.props.removeRow(this.state.id);
  }

  render() {
    const { t, actions, hasHandler } = this.props;
    return (
      <ItemDiv>
        {hasHandler && (
          <div className="dragHandler">
            <i className="ion-grid" />
          </div>
        )}
        <div className="flex3">
          <ItemDivInput
            name="description"
            type="text"
            value={this.state.description}
            onChange={this.handleTextInputChange}
            onKeyDown={this.handleKeyDown}
            placeholder={t('form:fields:items:description')}
          />
        </div>

        <div className="flex1">
          <ItemDivInput
            name="price"
            type="number"
            step="0.01"
            value={this.state.price}
            onChange={this.handleNumberInputChange}
            onKeyDown={this.handleKeyDown}
            placeholder={t('form:fields:items:price')}
          />
        </div>

        <div className="flex1">
          <ItemDivInput
            name="quantity"
            type="number"
            step="0.01"
            value={this.state.quantity}
            onChange={this.handleNumberInputChange}
            onKeyDown={this.handleKeyDown}
            placeholder={t('form:fields:items:quantity')}
          />
        </div>

        {(actions || hasHandler) && (
          <ItemActions>
            {actions && (
              <ItemRemoveBtn href="#" onClick={this.removeRow}>
                <i className="ion-close-circled" />
              </ItemRemoveBtn>
            )}
          </ItemActions>
        )}
      </ItemDiv>
    );
  }
}

ItemRow.propTypes = {
  actions: PropTypes.bool.isRequired,
  addItem: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
  hasHandler: PropTypes.bool.isRequired,
  item: PropTypes.object.isRequired,
  removeRow: PropTypes.func.isRequired,
  updateRow: PropTypes.func.isRequired,
};

export default compose(_withDraggable)(ItemRow);
