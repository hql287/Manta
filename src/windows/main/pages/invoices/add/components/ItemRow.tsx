// Libraries
import type React from 'react'
import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { Grip, Trash2 } from 'lucide-react'

import {
  Box,
  Grid,
  Flex,
  Group,
  Input,
  InputAddon,
  Stack,
  Text,
  Icon,
  IconButton,
} from '@chakra-ui/react'
import { NumberInputField, NumberInputRoot } from '@uiChakra/number-input'

interface ItemRowProps {
  actions: boolean
  addItem: () => void
  hasHandler: boolean
  item: unknown
  removeRow: (id: string) => void
  updateRow: (data: unknown) => void
}

// Component
export const ItemRow: React.FC<ItemRowProps> = ({
  actions,
  addItem,
  hasHandler,
  item,
  removeRow,
  updateRow,
}) => {
  const [state, setState] = useState({
    id: item.id,
    description: item.description || '',
    price: item.price || '',
    quantity: item.quantity || '',
    subtotal: item.subtotal || '',
  })

  useEffect(() => {
    updateSubtotal()
  }, [state.price, state.quantity])

  const handleKeyDown = (e) => {
    if (e.which === 13) {
      addItem()
    }
  }

  const handleTextInputChange = (event) => {
    const { name, value } = event.target
    setState((prevState) => ({ ...prevState, [name]: value }))
  }

  const handleNumberInputChange = (event) => {
    const { name, value } = event.target
    const parsedValue = value === '' ? '' : parseFloat(value)
    setState((prevState) => ({ ...prevState, [name]: parsedValue }))
  }

  const updateSubtotal = () => {
    const { price, quantity } = state
    const currentSubtotal =
      price && quantity ? parseFloat(price) * parseFloat(quantity) : ''
    setState((prevState) => ({ ...prevState, subtotal: currentSubtotal }))
    uploadRowState()
  }

  const uploadRowState = () => {
    updateRow(state)
  }

  const removeCurrentRow = () => {
    removeRow(state.id)
  }

  return (
    <Flex gap="4">
      {hasHandler && (
        <Flex alignItems="center">
          <Icon fontSize="sm" color="gray.300" size="sm">
            <Grip />
          </Icon>
        </Flex>
      )}
      <Grid templateColumns="3fr 120px 80px 1fr" gap="4" w="100%">
        <Input
          placeholder="Description"
          size="sm"
          onChange={handleTextInputChange}
          onKeyDown={handleKeyDown}
        />
        <Stack>
          <Group attached>
            <InputAddon>USD</InputAddon>
            <Input
              placeholder="99"
              size="sm"
              onChange={handleNumberInputChange}
              onKeyDown={handleKeyDown}
            />
          </Group>
        </Stack>

        <NumberInputRoot defaultValue="1" min={1} size="sm">
          <NumberInputField
            onChange={handleNumberInputChange}
            onKeyDown={handleKeyDown}
          />
        </NumberInputRoot>
        <Flex alignItems="center" justify="center">
          <Text fontSize="sm" color="gray.600" textAlign="end">
            Price
          </Text>
        </Flex>
      </Grid>

      {(actions || hasHandler) && (
        <Flex alignItems="center" justify="center" w="40px">
          {actions && (
            <IconButton
              colorPalette="red"
              variant="subtle"
              size="sm"
              aria-label="Search database"
              onClick={removeCurrentRow}
            >
              <Trash2 />
            </IconButton>
          )}
        </Flex>
      )}
    </Flex>
  )
}

export default ItemRow
