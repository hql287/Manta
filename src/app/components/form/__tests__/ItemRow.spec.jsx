// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

// Component
import { ItemRow } from '../ItemRow.jsx';

// Mocks
const actions = true;
const addItem = jest.fn();
const hasHandler = true;
const item = {
  id: '638fd7e3-217b-4a29-87f7-5ec539f73af9',
  description: 'iPhone X',
  price: 999,
  quantity: 1,
  subtotal: 999,
};
const removeRow = jest.fn();
const updateRow = jest.fn();
const t = jest.fn();

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <ItemRow
        t={t}
        actions={actions}
        addItem={addItem}
        hasHandler={hasHandler}
        item={item}
        removeRow={removeRow}
        updateRow={updateRow}
      />
    );
  });

  // PROPS & STATE
  it('receives correct props', () => {
    expect(wrapper.prop('actions')).toEqual(true);
    expect(wrapper.prop('actions')).not.toEqual(false);
    expect(wrapper.prop('hasHandler')).toEqual(true);
    expect(wrapper.prop('hasHandler')).not.toEqual(false);
    expect(wrapper.prop('item')).toEqual(item);
    expect(wrapper.prop('removeRow')).toEqual(removeRow);
    expect(wrapper.prop('updateRow')).toEqual(updateRow);
  });

  // RENDER
  it('renders necessary element', () => {
    expect(wrapper.find('input')).toHaveLength(3);
    expect(wrapper.find('a')).toHaveLength(1);
  });

  // LIFE CYCLE EVENTS
  it('set local state as item prop value', () => {
    expect(wrapper.state()).toEqual(item);
  });

  // PRIVATE METHOD
  it('handle text input change correctly', () => {
    // Setup
    const spy = jest.spyOn(ItemRow.prototype, 'handleTextInputChange');
    const wrap = mount(
      <ItemRow
        t={t}
        actions={actions}
        addItem={addItem}
        hasHandler={hasHandler}
        item={item}
        removeRow={removeRow}
        updateRow={updateRow}
      />
    );
    // Execute
    const textChange = {
      target: {
        name: 'description',
        value: 'iPhoneX',
      },
    };
    const descriptionInput = wrap.find('input').first();
    descriptionInput.simulate('change', textChange);
    // Assert
    expect(spy).toHaveBeenCalled();
  });

  it('handle number input change correctly', () => {
    const spy = jest.spyOn(ItemRow.prototype, 'handleNumberInputChange');
    const wrap = mount(
      <ItemRow
        t={t}
        actions={actions}
        addItem={addItem}
        hasHandler={hasHandler}
        item={item}
        removeRow={removeRow}
        updateRow={updateRow}
      />
    );

    // Execute
    const priceChange = {
      target: {
        name: 'price',
        value: '999',
      },
    };
    const priceInput = wrap.find('input').at(1);
    const quantityInput = wrap.find('input').last();
    priceInput.simulate('change', priceChange);
    // Assertion
    expect(spy).toHaveBeenCalled();
  });

  it('add new row when press enter', () => {
    const spy = jest.spyOn(ItemRow.prototype, 'handleKeyDown');
    const wrap = mount(
      <ItemRow
        t={t}
        actions={actions}
        addItem={addItem}
        hasHandler={hasHandler}
        item={item}
        removeRow={removeRow}
        updateRow={updateRow}
      />
    );
    // Execute
    const keyChange = {
      which: 13,
    };
    const descriptionInput = wrap.find('input').first();
    descriptionInput.simulate('keyDown', keyChange);
    // Assertion
    expect(spy).toHaveBeenCalled();
    expect(addItem).toHaveBeenCalled();
  });

  it('updateSubtotal when price and quantity change', () => {
    const spy = jest.spyOn(ItemRow.prototype, 'updateSubtotal');
    const wrap = mount(
      <ItemRow
        t={t}
        actions={actions}
        addItem={addItem}
        hasHandler={hasHandler}
        item={item}
        removeRow={removeRow}
        updateRow={updateRow}
      />
    );

    // Execute
    const priceChange = {
      target: {
        name: 'price',
        value: '799',
      },
    };
    const priceInput = wrap.find('input').at(1);
    priceInput.simulate('change', priceChange);
    // Assertion
    expect(spy).toHaveBeenCalled();
  });

  it('handle removeRow correctly', () => {
    const removeBtn = wrapper.find('a');
    removeBtn.simulate('click');
    expect(removeRow).toHaveBeenCalled();
  });

  // SNAPSHOT
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <ItemRow
          t={t}
          actions={actions}
          addItem={addItem}
          hasHandler={hasHandler}
          item={item}
          removeRow={removeRow}
          updateRow={updateRow}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
