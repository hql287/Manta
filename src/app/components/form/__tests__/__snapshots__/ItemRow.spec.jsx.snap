// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="ItemRow__ItemDiv-khgvaQ zZjdV"
>
  <div
    className="dragHandler"
  >
    <i
      className="ion-grid"
    />
  </div>
  <div
    className="flex3"
  >
    <input
      className="ItemRow__ItemDivInput-bHPGoF iXxPpt"
      name="description"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder={undefined}
      type="text"
      value="iPhone X"
    />
  </div>
  <div
    className="flex1"
  >
    <input
      className="ItemRow__ItemDivInput-bHPGoF iXxPpt"
      name="price"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder={undefined}
      step="0.01"
      type="number"
      value={999}
    />
  </div>
  <div
    className="flex1"
  >
    <input
      className="ItemRow__ItemDivInput-bHPGoF iXxPpt"
      name="quantity"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder={undefined}
      step="0.01"
      type="number"
      value={1}
    />
  </div>
  <div
    className="ItemRow__ItemActions-eRHgOi gPzpEq"
  >
    <a
      className="ItemRow__ItemRemoveBtn-kcypni jGmowk"
      href="#"
      onClick={[Function]}
    >
      <i
        className="ion-close-circled"
      />
    </a>
  </div>
</div>
`;
